### PR TITLE
Pool executions

### PR DESCRIPTION
**Pull request by AndrewBurian**
Less GC pressure if I'm not allocating a new struct, map, and array for each and every request.

The trade off is I have to iterate over the map and `delete` every entry from it before it can be reused.

Also updated the benchmarks to use `ServeHTTP` now that the code path is different between it and `Handler`

### CODE REVIEW

**Comment by chrisaxiom** [comment]
are there any benchmarks before/after?

File: execution_pool.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	return ep.p.Get().(*routeExecution)
+}
+
+func (ep *executionPool) Put(ex *routeExecution) {
+	ex.middleware = ex.middleware[0:0]
+	for key := range ex.params {
+		delete(ex.params, key)
+	}
+	ex.handler = nil
+	ex.notFound = nil
```
Comment (acroca) [inline execution_pool.go]: I think this is too coupled with the `routeExecution` model. What about adding a function like `ex.Clean()`, so the routeExecution will know how to clean everything up? I'm just worried some day we add new attributes to the `routeExecution` struct and we forget to come here and handle it...

**Comment by AndrewBurian** [comment]
Ugh. Benchmarks aren't super great. 

Before
```
BenchmarkSingleRoute-8      	 2000000	       692 ns/op	     528 B/op	       9 allocs/op
BenchmarkShallowAndWide-8   	 2000000	       960 ns/op	     528 B/op	       9 allocs/op
BenchmarkNarrowAndDeep-8    	  100000	     14699 ns/op	    2288 B/op	       9 allocs/op
BenchmarkFan-8              	 2000000	       941 ns/op	     528 B/op	       9 allocs/op
PASS
ok  	github.com/andrewburian/powermux	9.715s
 ```
After
```
benchmem -bench .
BenchmarkSingleRoute-8      	 2000000	       729 ns/op	     480 B/op	       8 allocs/op
BenchmarkShallowAndWide-8   	 1000000	      1009 ns/op	     480 B/op	       8 allocs/op
BenchmarkNarrowAndDeep-8    	  100000	     14632 ns/op	    2240 B/op	       8 allocs/op
BenchmarkFan-8              	 1000000	      1004 ns/op	     480 B/op	       8 allocs/op
```

Doesn't look like the allocs saved and GC time saved outweighs the cost of running the sync pool

**Comment by AndrewBurian** [comment]
Oh shii... didn't realize how much overhead there was on a defer call.

New After:
```
BenchmarkSingleRoute-8      	 2000000	       651 ns/op	     480 B/op	       8 allocs/op
BenchmarkShallowAndWide-8   	 2000000	       939 ns/op	     480 B/op	       8 allocs/op
BenchmarkNarrowAndDeep-8    	  100000	     14353 ns/op	    2240 B/op	       8 allocs/op
BenchmarkFan-8              	 2000000	       905 ns/op	     480 B/op	       8 allocs/op
PASS
ok  	github.com/andrewburian/powermux	9.411s
```

The savings are still pretty slim though. Still not sure if this is worth pushing through.

**Comment by acroca** [comment]
Performance wise the pool doesn't seem to help much, it'll for sure reduce some gc time and keep the memory less fragmented, but I don't think the introduced complexity (and risk of not doing a proper clean up when putting things back to the pool) justifies the feature.

**Comment by AndrewBurian** [comment]
Drove it down a bit more reusing `pathParts` slices. Saved a lot of memory on the worst case routes, but somehow still no fewer allocs.

```
BenchmarkSingleRoute-8      	 3000000	       633 ns/op	     480 B/op	       8 allocs/op
BenchmarkShallowAndWide-8   	 2000000	       866 ns/op	     480 B/op	       8 allocs/op
BenchmarkNarrowAndDeep-8    	  200000	     10581 ns/op	     481 B/op	       8 allocs/op
BenchmarkFan-8              	 2000000	       798 ns/op	     480 B/op	       8 allocs/op
PASS
ok  	github.com/andrewburian/powermux	10.047s
```

**Comment by AndrewBurian** [comment]
A few more tricks and we're now saving far more substantially on allocs and run time

```
BenchmarkSingleRoute-8      	 3000000	       443 ns/op	     400 B/op	       5 allocs/op
BenchmarkShallowAndWide-8   	 2000000	       717 ns/op	     400 B/op	       5 allocs/op
BenchmarkNarrowAndDeep-8    	  200000	     10120 ns/op	     401 B/op	       5 allocs/op
BenchmarkFan-8              	 2000000	       645 ns/op	     400 B/op	       5 allocs/op
PASS
ok  	github.com/andrewburian/powermux	8.305s
```

**Comment by turbolent** [comment]
I'm with @acroca here: introducing pools adds quite some complexity compared to very little performance gain, making the library harder to understand and maintain. 

Nice job though on the other refactoring and small optimizations, like duplicate `EscapedURL` calls.

I'd keep it simple and not have the pools, but 👍 for the optimizations.

**Comment by AndrewBurian** [comment]
I tested it with just the pools taken out, and it adds between 10% and 16% back on to the runtime. It also disproportionately hurts the kind of routes that could really use the most help, which are the ones with long paths or high number of path parameters. 14000ns down to 10000ns for the long path is quite an improvement.

In terms of code readability, the two pool `Get`/`Put` blocks are each contained to a single short function so you can at a glance see it coming and going. Those functions don't ever handle errors or return early so there's a low risk of missing a `Put` later. The `routeExecution` struct needed a constructor anyways to DRY up the few places it was being created, so adding the pool around it and doing it all in one file doesn't make it too much more confusing in my opinion.

**Comment by AndrewBurian** [comment]
Rewrote the history to take out the benchmark rewrite since that has been cherry picked to master.
You can now directly compare the performance on master to the performance on this branch.

**Comment by chrisaxiom** [comment]
You are really really trying to sell this hard aren't you @AndrewBurian ? 😄

**Comment by AndrewBurian** [comment]
It's been a long time since I got my teeth into a good optimization project. And of all the components of web service, this is the one I'd argue the most for milking every cycle out of it.

Also blame @Kay-Zee for the cherry pick 😛

**Comment by turbolent** [comment]
OK, thanks for the clarification Andrew. Code looks clean, so let's get this merged

**Review by turbolent — APPROVED** [review]

**Comment by Kay-Zee** [comment]
While I agree that readability is important, I think that the code is still quite clean, and a 30% increase is non negligible. 

We've also done a lot to circumvent possible mistakes, and check for race-conditions, so if the savings really is 30%, then I think its worth keeping.

**Pull request merged**